design-proposals/monitoring-code-refactor/monitoring-code-refactor.md

```diff
@@ -0,0 +1,84 @@
+# Monitoring code refactor
+
+**Authors:** João Vilaça <jvilaca@redhat.com>
+
+## What
+
+This design doc is proposing a code refactor for the monitoring logic in all
+KubeVirt components.
+
+## Why
+
+It’s important to refactor the monitoring code to separate monitoring logic from
+business logic. It will make the codebase more modular, readable, and
+maintainable. It will reduce complexity, reduce the risk of introducing errors,
+and allow us to test each component independently.
+
+### Pitfalls of the current solution
+
+We are facing multiple challenges with the current solution:
+
+- It is hard to keep a mental model of how monitoring is implemented in all the
+different repositories and is difficult to maintain all of them in sync with new
+features
+
+- Hard to add new features to metrics and alerts, for example, because of how
+metrics are being created it is hard to add a new metric stability field to all
+of them since they are independently created in many different places
+
+- Complex to collect all existing metrics and alerts for automatic documentation
+
+- High levels of intertwined monitoring and business logic code
+
+### Goals
+
+- Have a centralized place for the creation of monitoring resources
+- Have a consistent monitoring package between all projects
+- Have a code struct that is easy to maintain and evolve
+- Move closer to the Kubernetes metric implementation style
+- Have a structure and tools to accurately generate monitoring documentation,
+lint metrics and alerts, and define allow-list, deny-list and opt-in metrics
+
+### Non-Goals
+
+- Update any metrics or alerts
+- Apply these ideas to events in this phase
+- Create a new component to handle all project monitoring
+
+## How
+
+- Create a new separate “monitoring” package, in the "internal" folder,
```

Comment:
Not all projects are using the `internal` folder/s.
But if the suggestion here is to do so, it would be useful to explain why it helps and how it serves the needs listed above.

Reply:
Per [Golang project layout standards](https://github.com/golang-standards/project-layout#internal), I think it makes sense to use `internal/` to hold the application-specific metrics and rules code. The most important thing is that every project uses the same directories. But whether it is `internal/monitoring`, `pkg/monitoring` or even only `monitoring/` is up for discussion